### PR TITLE
added r_min and r_max to the RDF & NeighborList classes

### LINKED ISSUE #821: Require r_min >= 0 at the C++ level.
Currently the C++ code in freud does not ensure that `r_min` is non-negative (>= 0) in several classes. Values of `r_min < 0` should raise an error in the constructor like this:

https://github.com/glotzerlab/freud/blob/eb2a97bac302ed45236f26e0a7e6a0346071c579/cpp/density/RDF.cc#L17-L28

This applies to classes like `RDF` ~and `CorrelationFunction`~ and other classes that deal with radial distances that cannot be negative. This also affects `NeighborQuery` related classes, like `NeighborQueryPerPointIterator` (and related query constructors) and the method `NeighborList.filter_r`.

The structure factor calculations in #660 and #820 have `k_min` values where this same rule should apply (but that can be resolved in those PRs before merging them).

Fixing this will require tests that check for invalid values and ensure that an error is raised.

## Unwanted behavior:
```python
>>> freud.density.RDF(100, 10, r_min=-1)
freud.density.RDF(bins=100, r_max=10.0, r_min=-1.0
```

## Desired behavior:
```python
>>> freud.density.RDF(100, 10, r_min=-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "density.pyx", line 615, in freud.density.RDF.__cinit__
ValueError: RDF requires r_min to be non-negative.
```

### PR DESCRIPTION
**Pull request by SyedZiaul**
## Description
This PR adds error messages in `RDF` and `NeighborList`  for negative `r_min` and `r_max` values, as well as cases where `r_max <= r_min`.

## Motivation and Context

Partly resolves: #821 

## How Has This Been Tested?
Added tests that ensure errors are raised as expected when negative values are provided, or when `r_max < r_min`

## Types of changes

- [x] New feature (non-breaking change which adds or improves functionality)

## Checklist:

- [ ] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

File: cpp/density/RDF.cc [hunk 1]
```diff
@@ -21,6 +21,10 @@ RDF::RDF(unsigned int bins, float r_max, float r_min, bool normalize)
     if (r_max <= 0)
     {
         throw std::invalid_argument("RDF requires r_max to be positive.");
+    }
+        if (r_min < 0)
+    {
+        throw std::invalid_argument("RDF requires r_min to be non-negetive.");
```
Comment (bdice) [inline cpp/density/RDF.cc]:
```suggestion
        throw std::invalid_argument("RDF requires r_min to be non-negative.");
```

File: cpp/locality/NeighborList.cc [hunk 2]
```diff
@@ -163,6 +163,20 @@ template unsigned int NeighborList::filter(bool*);
...
+        throw std::invalid_argument("NeighborList requires r_max to be positive.");
+    }
+        if (r_min < 0)
+    {
+        throw std::invalid_argument("NeighborList requires r_min to be non-negetive.");
+    }
+    if (r_max <= r_min)
+    {
+        throw std::invalid_argument("NeighborList requires that r_max must be greater than r_min.");
+    }
```
Comment (bdice) [inline cpp/locality/NeighborList.cc]:
```suggestion
    if (r_max <= 0)
    {
        throw std::invalid_argument("NeighborList.filter_r requires r_max to be positive.");
    }
    if (r_min < 0)
    {
        throw std::invalid_argument("NeighborList.filter_r requires r_min to be non-negetive.");
    }
    if (r_max <= r_min)
    {
        throw std::invalid_argument("NeighborList.filter_r requires that r_max must be greater than r_min.");
    }
```

**Review by bdice — CHANGES_REQUESTED** [review]
Hi @SyedZiaul, it looks like you made a new GitHub account? Please avoid making multiple pull requests to address the same issue. You already opened #822 and #824 -- should those pull requests be closed?

Please see my code suggestions for changes.

**Comment by SyedZiaul** [comment]
@bdice okay i will keep it in mine mind, for the next time. i will not opened multiple pull request for same issue.

**Comment by SyedZiaul** [comment]
@bdice i closed #824 and #822 pull request. if their is any changes to be made in this regards , you can tell me.

File: cpp/density/RDF.cc [hunk 3]
```diff
@@ -21,6 +21,10 @@ RDF::RDF(unsigned int bins, float r_max, float r_min, bool normalize)
     if (r_max <= 0)
     {
         throw std::invalid_argument("RDF requires r_max to be positive.");
+    }
+        if (r_min < 0)
```
Comment (bdice) [inline cpp/density/RDF.cc]: Please check the indentation style. I keep requesting the same changes on every review.
```suggestion
    if (r_min < 0)
```

File: cpp/locality/NeighborList.cc [hunk 4]
```diff
@@ -163,6 +163,20 @@ template unsigned int NeighborList::filter(bool*);
 
 unsigned int NeighborList::filter_r(float r_max, float r_min)
 {
+
```
Comment (bdice) [inline cpp/locality/NeighborList.cc]: Delete this empty newline.
```suggestion
```

File: cpp/locality/NeighborList.cc [hunk 5]
```diff
@@ -163,6 +163,20 @@ template unsigned int NeighborList::filter(bool*);
...
+    if (r_max <= 0)
+    {
+        throw std::invalid_argument("NeighborList.filter_r requires r_max to be positive.");
+    }
+        if (r_min < 0)
```
Comment (bdice) [inline cpp/locality/NeighborList.cc]:
```suggestion
    if (r_min < 0)
```

**Review by bdice — CHANGES_REQUESTED** [review]
These changes will need to be accompanied by tests in the Python code. Please see examples here:

https://github.com/glotzerlab/freud/blob/2d989e0744beed8bb4194c587492ed1e3440578e/tests/test_density_RDF.py#L66-L73

Similar tests will be needed for `NeighborList.filter_r` in this file; https://github.com/glotzerlab/freud/blob/master/tests/test_locality_NeighborList.py#L38

Please create a new test function called `test_filter_r_invalid_arguments` for your new code in that file.

**Review by SyedZiaul — COMMENTED** [review]
@bdice any other changes need to made ?

**Comment by bdice** [comment]
> 
> 
> @bdice any other changes need to made ?

Please see my previous comment about adding tests. https://github.com/glotzerlab/freud/pull/825#pullrequestreview-741128899

**Comment by SyedZiaul** [comment]
@bdice i added this test function
### def test_filter_r_invalid_arguments(self):
        # Make sure that invalid NeighborList.filter_r objects raise errors
        with pytest.raises(ValueError):
            freud.locality.NeighborList.filter_r(r_max=0, r_min=1)
        with pytest.raises(ValueError):
            freud.locality.NeighborList.filter_r(r_max=1, r_min=-1)
        with pytest.raises(ValueError):
            freud.locality.NeighborList.filter_r(r_max=1, r_min=2)

i am getting some error , plz check?

**Comment by bdice** [comment]
@SyedZiaul I pushed a couple changes to this PR to fix the tests. See commits 53dc730 and f5658b5. I also applied our pull request message template to the PR description. Please fill out the pull request description above (by editing the PR description message) and update the credits/changelog. You can add a section at the bottom of the credits with your name. Once the PR template is filled out and the credits and changelog are updated, this should be ready to merge.

Some additional work is needed to fully resolve #821. I would recommend making a separate PR for these changes.
- [ ] Throw errors for invalid `r_max` and `diameter` arguments in LocalDensity.cc (see comments on #824, #822)
  - [ ] r_max > 0; must be positive
  - [ ] diameter >= 0; must be non-negative
- [ ] Throw errors for invalid arguments in the constructor of `NeighborQueryPerPointIterator`: https://github.com/glotzerlab/freud/blob/7d5662e614019f2b06493bb0a86b1ad8933bf4f3/cpp/locality/NeighborQuery.h#L262-L268
- [ ] Add tests for the errors mentioned above.

**Review by bdice — CHANGES_REQUESTED** [review]
This PR looks mostly complete, except for the changelog / credits I mentioned in my comment above.

**Comment by vyasr** [comment]
@SyedZiaul would you be able to finish up this PR? It looks like all that's left for you to do is update the PR template description above and update the changelog and credits.

**Comment by Charlottez112** [comment]
This seems good to go. I'll finish up what's left

**Comment by Charlottez112** [comment]
I'm not sure how I should update the changelog, so I'll just write it here:

## Change log

### Added
  * Added error checking for `r_min`, `r_max` in `freud.density.RDF` and `freud.locality.NeighborList`

**Review by Charlottez112 — APPROVED** [review]
Thank you for the work!

**Comment by bdice** [comment]
@Charlottez112 I'll edit this file: https://github.com/glotzerlab/freud/blob/master/ChangeLog.md

Then we can merge!

**Review by bdice — APPROVED** [review]

**Pull request merged**